File: C++/2-Medium/00153._寻找旋转排序数组中的最小值/00153._寻找旋转排序数组中的最小值.cpp
#include <iostream>
#include <vector>
using namespace std;

// Function to find the minimum element in a rotated sorted array
int findMin(vector<int>& nums) {
    int left = 0, right = nums.size() - 1;
    while (left < right) {
        int mid = left + (right - left) / 2;
        if (nums[mid] > nums[right]) {
            // The minimum element is in the right part
            left = mid + 1;
        } else {
            // The minimum element is in the left part
            right = mid;
        }
    }
    // When left == right, we found the minimum element
    return nums[left];
}

int main() {
    // Example cases
    vector<int> nums1 = {3, 4, 5, 1, 2};
    vector<int> nums2 = {4, 5, 6, 7, 0, 1, 2};
    // Output the results
    cout << "The minimum element in the first array is: " << findMin(nums1) << endl;
    cout << "The minimum element in the second array is: " << findMin(nums2) << endl;
    return 0;
}

File: Python/2-Medium/00153._寻找旋转排序数组中的最小值/00153._寻找旋转排序数组中的最小值.py
from typing import List

def findMin(nums: List[int]) -> int:
    """
    寻找旋转排序数组中的最小值

    Args:
    nums (List[int]): 旋转排序的数组

    Returns:
    int: 数组中的最小值

    示例:
    >>> findMin([3,4,5,1,2])
    1
    >>> findMin([4,5,6,7,0,1,2])
    0
    """
    left, right = 0, len(nums) - 1
    while left < right:
        mid = (left + right) // 2
        # 如果中间元素小于最右边元素，最小值在左半部分
        if nums[mid] < nums[right]:
            right = mid
        else:
            # 最小值在右半部分
            left = mid + 1
    # 当左指针等于右指针时，找到最小值
    return nums[left]

# 测试用例
test_cases = [
    ([3, 4, 5, 1, 2], 1),
    ([4, 5, 6, 7, 0, 1, 2], 0),
    ([11, 13, 15, 17], 11)  # 未旋转的数组
]

# 运行测试用例，验证代码正确性
results = []
for nums, expected in test_cases:
    result = findMin(nums)
    results.append((nums, result == expected, result))

print(results)